<launch>

  <arg name="name" default=""/>
  <arg name="port" default="/dev/ttyUSB0"/>
  <arg name="frame_id" default="base_laser"/>
  <arg name="frequency" default="6.0"/>
  <arg name="inverted" default="true"/>
  <arg name="use_sensor_data_qos" default="true"/>
  <arg name="count_subs_period" default="1.0"/>
  <arg name="x" default="0.0"/>
  <arg name="y" default="0.0"/>
  <arg name="z" default="0.0"/>
  <arg name="roll" default="0.0"/>
  <arg name="pitch" default="0.0"/>
  <arg name="yaw" default="0.0"/>
  <arg name="parent_frame_id" default=""/>
  <arg name="activation" default="$(eval '2 if __import__(\'importlib.util\').util.find_spec(\'rclpy.event_handler\') else 0')"/>
  <let name="tf2_ros_old_api" value="$(eval '\'$(env ROS_DISTRO)\' == \'foxy\'')"/>
  <let name="has_parent_frame" value="$(eval '\'$(var parent_frame_id)\' != \'\'')"/>

  <group>
    
    <push-ros-namespace namespace="$(var name)"/>

    <node name="ydlidar_driver" pkg="ydlidar_ros2_driver" exec="ydlidar_ros2_driver_node_2" output  ="screen" args="--ros-args --log-level info">
      <param name="port" value="$(var port)"/>
      <param name="frame_id" value="$(var name)/$(var frame_id)"/>
      <param name="frequency" value="$(var frequency)"/>
      <param name="ignore_array" value=""/>
      <param name="baudrate" value="230400"/>
      <param name="lidar_type" value="1"/>
      <param name="device_type" value="0"/>
      <param name="sample_rate" value="4"/>
      <param name="intensity_bit" value="8"/>
      <param name="fixed_resolution" value="true"/>
      <param name="reversion" value="true"/>
      <param name="inverted" value="$(var inverted)"/>
      <param name="use_sensor_data_qos" value="$(var use_sensor_data_qos)"/>
      <param name="count_subs_period" value="$(var count_subs_period)"/>
      <param name="auto_reconnect" value="true"/>
      <param name="isSingleChannel" value="false"/>
      <param name="intensity" value="true"/>
      <param name="support_motor_dtr" value="false"/>
      <param name="angle_max" value="180.0"/>
      <param name="angle_min" value="-180.0"/>
      <param name="range_max" value="12.0"/>
      <param name="range_min" value="0.03"/>
      <param name="activation" value="$(var activation)"/>
    </node>
  
    <include file="$(find-pkg-share ydlidar_ros2_driver)/launch/model.launch.py">
      <arg name="model" value="tmini_pro"/>
      <arg name="name" value="$(var name)"/>
      <arg name="lidar_frame_id" value="$(var frame_id)"/>
    </include>
  
    <group if="$(var has_parent_frame)">
      <group if="$(var tf2_ros_old_api)">
        <node name="static_tf_pub_lidar" pkg="tf2_ros" exec="static_transform_publisher" args="$(var x) $(var y) $(var z) $(var roll) $(var pitch) $(var yaw) $(var parent_frame_id) $(var name)/base_link" output="screen"  />
      </group>
      <group unless="$(var tf2_ros_old_api)">
        <node name="static_tf_pub_lidar" pkg="tf2_ros" exec="static_transform_publisher" args="  --x $(var x) --y $(var y) --z $(var z) --roll $(var roll) --pitch $(var pitch) --yaw   $(var yaw) --frame-id $(var parent_frame_id) --child-frame-id $(var name)/base_link "   output="screen"  />
      </group>
    </group>
  </group>
</launch>
